src/translate.py

```diff
@@ -4582,38 +4667,28 @@ def global_decls(
                     # The size of the element type (not the size of the array type)
                     if element_type is None:
                         element_type = sym.type
+
+                    # If we don't know the type, we can't guess the array_dim
                     type_size = element_type.get_size_bytes()
-                    if not type_size:
-                        # If we don't know the type, we can't guess the array_dim
-                        pass
-                    elif type_size > max_data_size:
-                        # Uh-oh! The type is too big for our data. (not an array)
-                        comments.append(
-                            f"type too large by {type_size - max_data_size}"
-                        )
-                    else:
-                        assert type_size <= max_data_size
-                        # We might have an array here. Now look at the lower bound,
-                        # which we know must be included in the initializer.
-                        data_size = min_data_size
-                        if data_size % type_size != 0:
-                            # How many extra bytes do we need to add to `data_size`
-                            # to make it an exact multiple of `type_size`?
-                            extra_bytes = type_size - (data_size % type_size)
-                            if data_size + extra_bytes <= max_data_size:
-                                # We can make an exact multiple by taking some of the bytes
-                                # we thought were padding
-                                data_size += extra_bytes
-                            else:
-                                comments.append(f"extra bytes: {data_size % type_size}")
-                        if data_size // type_size > 1 or is_vla:
-                            # We know it's an array
-                            array_dim = max_data_size // type_size
+                    if type_size:
+                        potential_dim, extra_bytes = sym.potential_array_dim(type_size)
+                        if potential_dim == 0 and extra_bytes > 0:
+                            # The type is too big for our data. (not an array)
+                            comments.append(
+                                f"type too large by {fmt.format_int(type_size - extra_bytes)}"
```

Comment:
```suggestion
                                f"type too large by {fmt.format_int(extra_bytes)}"
```

Reply:
The original was right, though the refactor is confusing. I checked the diff to be sure:

```diff
-static GlobalContext D_801C0EAC;                    /* type too large by 102988; unable to generate initializer */
+static GlobalContext D_801C0EAC;                    /* type too large by 0x1924C; unable to generate initializer */
-static Player D_801C0EB8;                           /* type too large by 3440; unable to generate initializer */
+static Player D_801C0EB8;                           /* type too large by 0xD70; unable to generate initializer */
-static CollisionHeader D_80AB1690;                  /* type too large by 20; unable to generate initializer */
+static CollisionHeader D_80AB1690;                  /* type too large by 0x14; unable to generate initializer */
-static Actor D_8085D5E0;                            /* type too large by 320; unable to generate initializer */
+static Actor D_8085D5E0;                            /* type too large by 0x140; unable to generate initializer */
```

The return value from `potential_array_dim` is analogous to the return of `divmod`. `extra_bytes` represents the number of *unused* bytes: if `potential_dim  == 0`, then the data was only `extra_bytes` large but the type is `type_size`.